oracle/oracle.go

```diff
@@ -0,0 +1,377 @@
+package oracle
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/libopenstorage/cloudops"
+	"github.com/oracle/oci-go-sdk/v65/common"
+	"github.com/oracle/oci-go-sdk/v65/containerengine"
+	"github.com/oracle/oci-go-sdk/v65/core"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	v1MetadataAPIEndpoint = "http://169.254.169.254/opc/v1/instance/"
+	v2MetadataAPIEndpoint = "http://169.254.169.254/opc/v2/instance/"
+	metadataInstanceIDkey = "id"
+	// MetadataRegionKey key name for region in metadata JSON
+	// returned by IMDS service
+	MetadataRegionKey = "canonicalRegionName"
+	// MetadataAvailabilityDomainKey key name for availability domain
+	// in metadata JSON returned by IMDS service
+	MetadataAvailabilityDomainKey = "availabilityDomain"
+	metadataCompartmentIDkey      = "compartmentId"
+	// MetadataKey key name in metadata json for metadata returned by IMDS service
+	MetadataKey = "metadata"
+	// MetadataUserDataKey key name in metadata json for user data
+	MetadataUserDataKey   = "user_data"
+	metadataTenancyIDKey  = "oke-tenancy-id"
+	metadataPoolIDKey     = "oke-pool-id"
+	envPrefix             = "PX_ORACLE"
+	envInstanceID         = "INSTANCE_ID"
+	envRegion             = "INSTANCE_REGION"
+	envAvailabilityDomain = "INSTNACE_AVAILABILITY_DOMAIN"
+	envCompartmentID      = "COMPARTMENT_ID"
+	envTenancyID          = "TENANCY_ID"
+	envPoolID             = "POOL_ID"
+)
+
+type oracleOps struct {
+	cloudops.Compute
+	cloudops.Storage
+	instance           string
+	region             string
+	availabilityDomain string
+	compartmentID      string
+	tenancyID          string
+	poolID             string
+	storage            core.BlockstorageClient
+	compute            core.ComputeClient
+	containerEngine    containerengine.ContainerEngineClient
+}
+
+// NewClient creates a new cloud operations client for Oracle cloud
+func NewClient() (cloudops.Ops, error) {
+	oracleOps := &oracleOps{}
+	err := getInfoFromMetadata(oracleOps)
+	if err != nil {
+		fmt.Printf("Got error [%v] from metadata\n", err)
+		err = getInfoFromEnv(oracleOps)
+		if err != nil {
+			return nil, err
+		}
+	}
+	os.Setenv(fmt.Sprintf("%s_tenancy_ocid", envPrefix), oracleOps.tenancyID)
+	os.Setenv(fmt.Sprintf("%s_region", envPrefix), oracleOps.region)
+	configProvider := common.ConfigurationProviderEnvironmentVariables(envPrefix, "")
+	oracleOps.storage, err = core.NewBlockstorageClientWithConfigurationProvider(configProvider)
+	if err != nil {
+		return nil, err
+	}
+	oracleOps.compute, err = core.NewComputeClientWithConfigurationProvider(configProvider)
+	if err != nil {
+		return nil, err
+	}
+	oracleOps.containerEngine, err = containerengine.NewContainerEngineClientWithConfigurationProvider(configProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: wrap around exponentialBackoffOps
```

Comment:
Do we have a ticket tracking this?

Reply:
Filed https://portworx.atlassian.net/browse/PWX-26214